src/Interpreters/CancellationChecker.cpp

```diff
@@ -0,0 +1,139 @@
+#include <Common/logger_useful.h>
+#include "Interpreters/ProcessList.h"
+#include <Interpreters/CancellationChecker.h>
+
+#include <__chrono/duration.h>
+#include <condition_variable>
+#include <mutex>
+
+
+namespace DB
+{
+
+QueryToTrack::QueryToTrack(
+    std::shared_ptr<QueryStatus> query_,
+    UInt64 timeout_,
+    UInt64 endtime_)
+    : query(query_), timeout(timeout_), endtime(endtime_)
+{
+}
+
+bool CompareEndTime::operator()(const QueryToTrack& a, const QueryToTrack& b) const
+{
+    if (a.endtime != b.endtime)
+        return a.endtime < b.endtime;
+    else
+        return a.query < b.query;
+}
+
+CancellationChecker::CancellationChecker() : stop_thread(false)
+{
+}
+
+CancellationChecker::~CancellationChecker()
+{
+    stop_thread = true;
```

Comment:
It is useless. You need to make designated method which sets `stop_thread` as `true`. It has to be used right before the `cancellation_task.join()` call.